### PR TITLE
Wait for proxy to propagated to cluster

### PR DESCRIPTION
**Pull request by praveenkumar**
Since we are not disabling the CVO so in case of proxy, CVO
take more time to propagate it to different operators, at this
moment we are checking operator status in every 2 sec (60 times)
till all are available.

### CODE REVIEW

File: pkg/crc/machine/machine.go [hunk 1]
```diff
@@ -415,6 +415,11 @@ func Start(startConfig StartConfig) (StartResult, error) {
 		return *result, errors.Newf("Error approving the node csr %v", err)
 	}
 
+	if proxyConfig.IsEnabled() {
+		logging.Info("Adding proxy details to cluster ...")
```
Comment (cfergeau) [inline pkg/crc/machine/machine.go]: "Waiting for the proxy configuration to be applied" or something like this? "Adding" sounds like we are doing an action on th ecluster when we are only waiting for things to happen :)

File: pkg/crc/machine/machine.go [hunk 2]
```diff
@@ -830,3 +835,22 @@ func configProxyForCluster(ocConfig oc.OcConfig, sshRunner *crcssh.SSHRunner, sd
...
+		}
+		if !status {
+			logging.Debug("Proxy changes for cluster in progress")
+			return &errors.RetriableError{Err: fmt.Errorf("all operators are still not in available state")}
+		}
+		return nil
+	}
+
+	if err := errors.RetryAfter(60, clusterstatus, 2*time.Second); err != nil {
+		logging.Debug("Cluster is still in progress state after proxy is applied.")
```
Comment (cfergeau) [inline pkg/crc/machine/machine.go]: "Proxy settings have not been applied yet"?

File: pkg/crc/machine/machine.go [hunk 3]
```diff
@@ -830,3 +835,22 @@ func configProxyForCluster(ocConfig oc.OcConfig, sshRunner *crcssh.SSHRunner, sd
...
+func checkProxyPropagatedToCluster(ocConfig oc.OcConfig, proxyConfig *network.ProxyConfig) {
+	clusterstatus := func() error {
+		status, err := cluster.CheckProxyIsPartOfAuthOperator(ocConfig, proxyConfig)
+		if err != nil {
+			logging.Debugf("Error getting the clusterstatus %v", err)
+			return &errors.RetriableError{Err: err}
+		}
+		if !status {
+			logging.Debug("Proxy changes for cluster in progress")
+			return &errors.RetriableError{Err: fmt.Errorf("all operators are still not in available state")}
```
Comment (cfergeau) [inline pkg/crc/machine/machine.go]: Is this the right error message or does this need to be adjusted?

File: pkg/crc/machine/machine.go [hunk 4]
```diff
@@ -830,3 +835,22 @@ func configProxyForCluster(ocConfig oc.OcConfig, sshRunner *crcssh.SSHRunner, sd
 
 	return nil
 }
+
+func checkProxyPropagatedToCluster(ocConfig oc.OcConfig, proxyConfig *network.ProxyConfig) {
```
Comment (cfergeau) [inline pkg/crc/machine/machine.go]: waitForProxyPropagation?

File: pkg/crc/machine/machine.go [hunk 5]
```diff
@@ -830,3 +835,22 @@ func configProxyForCluster(ocConfig oc.OcConfig, sshRunner *crcssh.SSHRunner, sd
...
+	clusterstatus := func() error {
+		status, err := cluster.CheckProxyIsPartOfAuthOperator(ocConfig, proxyConfig)
+		if err != nil {
+			logging.Debugf("Error getting the clusterstatus %v", err)
```
Comment (cfergeau) [inline pkg/crc/machine/machine.go]: "clusterstatus" does not sound correct?

File: pkg/crc/cluster/cluster.go [hunk 6]
```diff
@@ -223,3 +223,15 @@ func DeleteOpenshiftApiServerPods(oc oc.OcConfig) error {
 
 	return errors.RetryAfter(60, deleteOpenshiftApiserverPods, time.Second)
 }
+
+func CheckProxyIsPartOfAuthOperator(oc oc.OcConfig, proxy *network.ProxyConfig) (bool, error) {
```
Comment (cfergeau) [inline pkg/crc/cluster/cluster.go]: `AuthOperator`? this is checking `openshift-marketplace`? I would pass the operator name as an argument to that method, and call it `CheckProxySettingsForOperator(...)`, `OperatorHasProxy`, ... or something like this?

File: pkg/crc/cluster/cluster.go [hunk 7]
```diff
@@ -223,3 +223,15 @@ func DeleteOpenshiftApiServerPods(oc oc.OcConfig) error {
...
+	cmdArgs := []string{"set", "env", "deployment", "redhat-operators", "--list", "-n", "openshift-marketplace"}
+	out, _, err := oc.RunOcCommandPrivate(cmdArgs...)
+	if err != nil {
+		return false, err
+	}
+	if strings.Contains(out, proxy.HttpsProxy) {
```
Comment (cfergeau) [inline pkg/crc/cluster/cluster.go]: What if only HttpProxy is set?

File: pkg/crc/machine/machine.go [hunk 8]
```diff
@@ -830,3 +835,22 @@ func configProxyForCluster(ocConfig oc.OcConfig, sshRunner *crcssh.SSHRunner, sd
 
 	return nil
 }
+
+func checkProxyPropagatedToCluster(ocConfig oc.OcConfig, proxyConfig *network.ProxyConfig) {
+	clusterstatus := func() error {
+		status, err := cluster.CheckProxyIsPartOfAuthOperator(ocConfig, proxyConfig)
```
Comment (cfergeau) [inline pkg/crc/machine/machine.go]: `status` is not really accurate, `proxySet` maybe?

File: pkg/crc/cluster/cluster.go [hunk 9]
```diff
@@ -223,3 +223,15 @@ func DeleteOpenshiftApiServerPods(oc oc.OcConfig) error {
 
 	return errors.RetryAfter(60, deleteOpenshiftApiserverPods, time.Second)
 }
+
+func CheckProxyIsPartOfAuthOperator(oc oc.OcConfig, proxy *network.ProxyConfig) (bool, error) {
+	cmdArgs := []string{"set", "env", "deployment", "redhat-operators", "--list", "-n", "openshift-marketplace"}
```
Comment (cfergeau) [inline pkg/crc/cluster/cluster.go]: Maybe it's worth starting the method with `if !proxy.IsEnabled() { return true, nil }`, potentially with a `logging.Debugf` to know what's going on?

File: pkg/crc/machine/machine.go (hunk 1 above)
Comment (praveenkumar) [reply pkg/crc/machine/machine.go]: @cfergeau since here we are already check for `IsEnabled()` does it make sense to check https://github.com/code-ready/crc/pull/1072/files#r388997725 here?

File: pkg/crc/cluster/cluster.go (hunk 7 above)
Comment (praveenkumar) [reply pkg/crc/cluster/cluster.go]: Ah yes I have to put an or operator here, had in mind but just missed when putting the code :(

File: pkg/crc/cluster/cluster.go (hunk 6 above)
Comment (praveenkumar) [reply pkg/crc/cluster/cluster.go]: Ah I was testing Auth before but yes I will pass the namespace and deployment of operator as argument.

File: pkg/crc/machine/machine.go (hunk 1 above)
Comment (cfergeau) [reply pkg/crc/machine/machine.go]: We are adding a new exported `CheckProxyIsPartOfAuthOperator` method in the `cluster` package. It's currently only called from `machine.go`, but since it's part of a different package, maybe one day there will be more users calling it. Either we consider it's an error to call `CheckProxyIsPartOfAuthOperator` if `proxy.IsEnabled()` is false, in which case it would still be nice for `CheckProxyIsPartOfAuthOperator` to have a warning about it, or we consider it's ok to call `CheckProxyIsPartOfAuthOperator` if `proxy.IsEnabled()` is false, and then it should return true. In any case, I don't think we should rely on the rest of the code to have properly checked that proxy is enabled before calling that method.

File: pkg/crc/cluster/cluster.go [hunk 10]
```diff
@@ -223,3 +223,15 @@ func DeleteOpenshiftApiServerPods(oc oc.OcConfig) error {
 
 	return errors.RetryAfter(60, deleteOpenshiftApiserverPods, time.Second)
 }
+
+func CheckProxySettingsForOperator(oc oc.OcConfig, proxy *network.ProxyConfig, deployment, namespace string) (bool, error) {
+	cmdArgs := []string{"set", "env", "deployment", deployment, "--list", "-n", namespace}
+	out, _, err := oc.RunOcCommandPrivate(cmdArgs...)
```
Comment (cfergeau) [inline pkg/crc/cluster/cluster.go]: why `Private`?

File: pkg/crc/machine/machine.go [hunk 11]
```diff
@@ -830,3 +835,22 @@ func configProxyForCluster(ocConfig oc.OcConfig, sshRunner *crcssh.SSHRunner, sd
...
+		}
+		if !proxySet {
+			logging.Debug("Proxy changes for cluster in progress")
+			return &errors.RetriableError{Err: fmt.Errorf("")}
+		}
+		return nil
+	}
+
+	if err := errors.RetryAfter(60, checkProxySettingsForOperator, 2*time.Second); err != nil {
+		logging.Debug("Proxy settings have not been applied yet to cluster")
```
Comment (cfergeau) [inline pkg/crc/machine/machine.go]: Sounds a bit like "the settings are not applied yet, but they are going to be applied later", which we don't know as we just failed to apply them. I would reword this to be less optismistic "Failed to propagate proxy settings to cluster" or something like this.

File: pkg/crc/cluster/cluster.go (hunk 10 above)
Comment (praveenkumar) [reply pkg/crc/cluster/cluster.go]: @cfergeau as part of output it list the proxy details.

**Comment by praveenkumar** [comment]
retest this please
